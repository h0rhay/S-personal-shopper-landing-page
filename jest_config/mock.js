export const mockTheme = {
  themeType: 'dark',
  breakpoints: {
    small: '(max-width: 812px)',
    medium: '(min-width: 813px)',
    large: '(min-width: 960px)',
    deviceSwitchSm: '(max-width: 896px)',
    deviceSwitchLg: '(min-width: 897px)',
  },
  typography: {
    mobile: {
      headerOne: {
        fontSize: '3rem',
      },
      headerTwo: {
        fontSize: '1.75rem',
      },
      displayHeading: {
        fontSize: 'calc(1.5rem * 2)',
      },
      itemHeading: {
        fontSize: '1.125rem',
      },
    },
    desktop: {
      headerOne: {
        fontSize: '4rem',
      },
      headerTwo: {
        fontSize: '2rem',
      },
      displayHeading: {
        fontSize: 'calc(4rem * 2)',
      },
      itemHeading: {
        fontSize: '1.125rem',
      },
    },
  },
  text: '#ffffff',
  primary: '#212121',
  primaryAccent: '#2E2E2E',
  accent: '#ffe256',
  shadow: 'rgba(0, 0, 0, 0.4)',
};
