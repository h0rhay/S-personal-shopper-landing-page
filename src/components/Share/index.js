import React, { useState, useEffect } from 'react';
import {
  FacebookShareButton,
  FacebookIcon,
  TwitterShareButton,
  TwitterIcon,
  LinkedinShareButton,
  LinkedinIcon,
  WhatsappShareButton,
  WhatsappIcon,
} from 'react-share';
import styled from 'styled-components';
import ShareBadge from '../ShareBadge';
import Modal from '../Modal';
import CopyIconSVG from '../../images/copy_icon.svg';
import CopiedIconSVG from '../../images/tick_icon.svg';

const SocialNetworkingWrapper = styled.div`
  display: flex;
  flex-wrap: wrap;
  justify-content: center;
  margin-bottom: 1.5rem;
  text-align: center;
`;

const SocialNetworkingButtonWrapper = styled.div`
  margin: 0.5rem 1rem;
  display: flex;
  flex-direction: column;
  justify-content: flex-end;
  span {
    color: #212121;
    font-size: 0.7rem;
  }
`;

const CopyIcon = styled(CopyIconSVG)`
  height: 3rem;
  width: 3rem;
  cursor: pointer;
`;

const CopiedIcon = styled(CopiedIconSVG)`
  height: 3.2rem;
  width: 3.2rem;
`;

const CopyUrlbutton = styled.button`
  background: transparent;
  border: none;
`;

const ShareHeader = styled.h3`
  color: #212121;
  text-align: center;
  margin-top: 2rem;
  letter-spacing: 2px;
`;

const Share = () => {
  const [showDialog, setShowDialog] = useState(false);
  const [copyStatus, setCopyStatus] = useState(false);

  useEffect(() => {
    setTimeout(() => {
      if (copyStatus) setCopyStatus(false);
    }, 5000);
  }, [copyStatus]);

  const shareDetails = {
    title: 'Selfridges Personal Shopper',
    subject: 'Personal shopping at Selfridges',
    body: 'Check out my favourite Selfridges pieces!',
    url: document.location.href,
    hashtag: ['personalshopping', 'selfridges'],
  };

  const copyToClipboard = () => {
    navigator.clipboard.writeText(shareDetails.url);
    setCopyStatus(true);
  };

  const openSharePopup = () => {
    if (navigator.share) {
      navigator
        .share({
          title: shareDetails.title,
          text: shareDetails.body,
          url: shareDetails.url,
        })
        .then(() => {})
        .catch(() => {});
    } else {
      setCopyStatus(false);
      setShowDialog(true);
    }
    return null;
  };

  const closePopup = () => setShowDialog(false);

  return (
    <div>
      <ShareBadge clickHandler={openSharePopup} />

      <Modal isOpen={showDialog} onDismiss={closePopup}>
        <ShareHeader>Share this page</ShareHeader>
        <SocialNetworkingWrapper>
          <SocialNetworkingButtonWrapper>
            <WhatsappShareButton
              title={shareDetails.title}
              url={shareDetails.url}
              summary={shareDetails.body}
              source={shareDetails.url}
            >
              <WhatsappIcon size={50} round />
            </WhatsappShareButton>
            <span>Whatsapp</span>
          </SocialNetworkingButtonWrapper>
          <SocialNetworkingButtonWrapper>
            <FacebookShareButton
              url={shareDetails.url}
              quote={shareDetails.subject}
              hashtag={`#${shareDetails.hashtag[0]}`}
              description={shareDetails.body}
            >
              <FacebookIcon size={50} round />
            </FacebookShareButton>
            <span>Facebook</span>
          </SocialNetworkingButtonWrapper>
          <SocialNetworkingButtonWrapper>
            <TwitterShareButton title={shareDetails.body} url={shareDetails.url} hashtags={shareDetails.hashtag}>
              <TwitterIcon size={50} round />
            </TwitterShareButton>
            <span>Twitter</span>
          </SocialNetworkingButtonWrapper>
          <SocialNetworkingButtonWrapper>
            <LinkedinShareButton
              title={shareDetails.title}
              url={shareDetails.url}
              summary={shareDetails.body}
              source={shareDetails.url}
            >
              <LinkedinIcon size={50} round />
            </LinkedinShareButton>
            <span>Linkedin</span>
          </SocialNetworkingButtonWrapper>
          <SocialNetworkingButtonWrapper>
            <CopyUrlbutton data-testid="copy-link-button" onClick={copyToClipboard}>
              {copyStatus ? <CopiedIcon /> : <CopyIcon />}
            </CopyUrlbutton>
            <span>{copyStatus ? 'Link copied' : 'Copy link'}</span>
          </SocialNetworkingButtonWrapper>
        </SocialNetworkingWrapper>
      </Modal>
    </div>
  );
};

export default Share;
