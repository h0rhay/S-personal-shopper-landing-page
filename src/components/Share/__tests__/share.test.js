import React from 'react';
import { render, screen, fireEvent, act } from '@testing-library/react';
import { ThemeProvider } from 'styled-components';
import { mockTheme } from '../../../../jest_config/mock';
import Share from '../index';

let container;
const renderComponent = ({ theme }) =>
  render(
    <ThemeProvider theme={theme}>
      <Share />
    </ThemeProvider>
  );

describe('Share component', () => {
  beforeEach(() => {
    container = renderComponent({ theme: mockTheme }).container;
  });

  it('Should match screenshot', () => {
    expect(container).toMatchSnapshot();
  });

  describe('When share button is clicked', () => {
    beforeEach(() => {
      fireEvent.click(screen.getByRole('button', { name: /share/i }));
      Object.assign(window.navigator, {
        clipboard: {
          writeText: jest.fn().mockImplementation(() => Promise.resolve()),
        },
      });
      jest.useFakeTimers();
    });
    it('Modal should appear', () => {
      const modal = screen.getByText(/share this page/i);
      expect(modal).toBeInTheDocument();
      expect(modal).toBeInTheDocument();
    });

    it('Whatsapp share button should appear', () => {
      const whatsapp = screen.getByText(/whatsapp/i);
      expect(whatsapp).toBeInTheDocument();
    });

    it('Facebook share button should appear', () => {
      const facebook = screen.getByText(/facebook/i);
      expect(facebook).toBeInTheDocument();
    });

    it('Twitter share button should appear', () => {
      const twitter = screen.getByText(/twitter/i);
      expect(twitter).toBeInTheDocument();
    });

    it('Linkedin share button should appear', () => {
      const linkedin = screen.getByText(/linkedin/i);
      expect(linkedin).toBeInTheDocument();
    });

    it('Copy Link button should appear', () => {
      expect(screen.getByText(/copy link/i)).toBeInTheDocument();

      act(() => {
        const copyLinkButton = screen.getByTestId('copy-link-button');
        fireEvent.click(copyLinkButton);
      });

      expect(screen.getByText(/link copied/i)).toBeInTheDocument();
      expect(screen.queryByText(/copy link/i)).not.toBeInTheDocument();

      act(() => jest.advanceTimersByTime(5000));

      expect(screen.getByText(/copy link/i)).toBeInTheDocument();
      expect(screen.queryByText(/link copied/i)).not.toBeInTheDocument();
    });
  });
});
