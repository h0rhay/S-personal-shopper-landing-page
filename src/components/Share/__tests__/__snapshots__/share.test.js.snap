// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Share component Should match screenshot 1`] = `
<div>
  <div>
    <button
      class="sc-bdvvtL kOTLLx"
      role="button"
      type="button"
    >
      <div>
        <test-file-stub />
        <div>
          SHARE
        </div>
      </div>
    </button>
    <div />
  </div>
</div>
`;
